Handle windows popping above the slimlock better.

aartamonau opened the pull request:
Certain windows (especially notifications like notify-osd, etc) tend
to make their windows "override redirect". Such windows can't be
prevented from popping on top of slimlock. Previous approach of
handling this was to raise our window every second. This is definitely
better than nothing but potentially sensitive information is still
revealed for some time. And actually it didn't work at least for me
anyway: for some reason XRaiseWindow would not send Expose
event. Probably it has something to do with the fact that this
XRaiseWindow is send from another thread and Xlib is infamous for its
multithreading support. So the new approach is to handle
VisibilityNotify events and raise the window in case the event states
that our window is partially or fully obscured. This is still not
perfect since popping window will be seen as noticeable artifact. But
it seems that x11 does not give a way to handle it better.

aartamonau commented:
I just realized that it's not going to work with compositing enabled. So I should probably add a periodic call to XRaiseWindow as a fallback for such a case.

aartamonau commented:
Added another commit to raise window in the old way. Which is needed as I said above when compositing is enabled.

Commit:
Clean trailing whitespaces

Commit:
Unlock lockfile before close it

dannyn committed:
Fixed crtc bug.

dannyn committed:
Background across multiple monitors.

dannyn committed:
Merge branch 'master' of https://github.com/renatoaguiar/slimlock

AlexJF committed:
Fixed early free

AlexJF committed:
Merge branch 'master' of git://github.com/dannyn/slimlock

dannyn committed:
Fixed keyboard grabbing issue.

dannyn committed:
Merge pull request #1 from AlexJF/master

Fix early free

bentglasstube committed:
Add config option to set password mask character

Commit:
added uninstall section to Makefile

Commit:
code cleanup

Commit:
reverted changes; compile error

dannyn committed:
Merge pull request #2 from bentglasstube/password-char

Add config option to set password mask character

dannyn committed:
Merge pull request #3 from Unia/master

Add uninstall section to Makefile and several code cleanups

wapiflapi commented:
Hi,

I know this has been made over a year ago, but as it is still open I thought I'dd my two cents here.

@joelburget is no longer maintaining this project. Maybe it would be a good idea to submit this pull request to @dannyn /slimlock ?

aartamonau committed:
Add a configurable way to raise slimlock window periodically.

This change supplements a previous attempt to handle popping windows
better. I realized that approach based on mere handling of visibility
events doesn't work when compositing is enabled. Thus it's still
needed to raise slimlock window from time to time. So I added
`raise_interval' configuration option which specifies how often the
window must be raised (default value is every 1000ms). It's possible
to disable raising completely by setting `raise_interval' to
zero. Everything is implemented in a single thread to avoid notorious
Xlib multi-threading issues.

aartamonau committed:
Handle windows popping above the slimlock better.

Certain windows (especially notifications like notify-osd, etc) tend
to make their windows "override redirect". Such windows can't be
prevented from popping on top of slimlock. Previous approach of
handling this was to raise our window every second. This is definitely
better than nothing but potentially sensitive information is still
revealed for some time. And it didn't work at least for me anyway: for
some reason XRaiseWindow would not send Expose event. Probably it has
something to do with the fact that this XRaiseWindow is send from
another thread and Xlib is infamous for its multithreading support. So
the new approach is to handle VisibilityNotify events and raise the
window in case the event states that our window is partially or fully
obscured. This is still not perfect since popping window will be seen
as noticeable artifact. But it seems that x11 does not give a way to
handle it better.
